File: backend/seed.py
"""
Seed script: Creates a default tenant, admin user, sample departments, projects,
and demo usage data so the dashboard is populated for testing.

Usage:
  cd backend
  python seed.py
"""
import asyncio
import uuid
import random
from datetime import date, timedelta, datetime, timezone
from decimal import Decimal

from sqlalchemy import select

from app.database import engine, async_session, Base
from app.models import *  # noqa
from app.services.auth_service import hash_password
from app.services.encryption_service import encrypt_api_key

MODELS_BY_PLATFORM = {
    "openai": ["gpt-4o", "gpt-4o-mini", "gpt-4-turbo", "gpt-3.5-turbo", "text-embedding-3-small"],
    "anthropic": ["claude-sonnet-4-20250514", "claude-3.5-haiku", "claude-3-opus"],
    "google": ["gemini-2.0-flash", "gemini-1.5-pro", "gemini-1.5-flash"],
}


async def seed():
    async with engine.begin() as conn:
        await conn.run_sync(Base.metadata.create_all)

    async with async_session() as db:
        # Check if already seeded
        existing = await db.execute(select(Tenant))
        if existing.scalar_one_or_none():
            print("Database already seeded. Skipping.")
            return

        # Tenant
        tenant = Tenant(name="Demo Organization", slug="demo-org")
        db.add(tenant)
        await db.flush()

        # Departments
        eng_dept = Department(tenant_id=tenant.id, name="Engineering", description="Software engineering team")
        ml_dept = Department(tenant_id=tenant.id, name="Machine Learning", description="ML/AI research team")
        product_dept = Department(tenant_id=tenant.id, name="Product", description="Product management")
        db.add_all([eng_dept, ml_dept, product_dept])
        await db.flush()

        # Admin user
        admin = User(
            tenant_id=tenant.id, department_id=eng_dept.id,
            email="admin@demo.com", password_hash=hash_password("admin123"),
            full_name="Admin User", role="admin",
        )
        # Regular users
        users = [
            User(tenant_id=tenant.id, department_id=eng_dept.id, email="alice@demo.com", password_hash=hash_password("user123"), full_name="Alice Engineer", role="user"),
            User(tenant_id=tenant.id, department_id=ml_dept.id, email="bob@demo.com", password_hash=hash_password("user123"), full_name="Bob Researcher", role="user"),
            User(tenant_id=tenant.id, department_id=product_dept.id, email="carol@demo.com", password_hash=hash_password("user123"), full_name="Carol PM", role="user"),
        ]
        db.add(admin)
        db.add_all(users)
        await db.flush()

        # Projects
        proj_chatbot = Project(tenant_id=tenant.id, department_id=eng_dept.id, name="Chatbot Production", description="Customer-facing chatbot")
        proj_internal = Project(tenant_id=tenant.id, department_id=eng_dept.id, name="Internal Tools", description="Internal dev tools using AI")
        proj_research = Project(tenant_id=tenant.id, department_id=ml_dept.id, name="ML Research", description="ML experimentation")
        proj_content = Project(tenant_id=tenant.id, department_id=product_dept.id, name="Content Generation", description="Marketing content generation")
        db.add_all([proj_chatbot, proj_internal, proj_research, proj_content])
        await db.flush()

        # User-project access
        for u in users:
            db.add(UserProjectAccess(user_id=u.id, project_id=proj_chatbot.id, permission_level="view"))

        db.add(UserProjectAccess(user_id=users[0].id, project_id=proj_internal.id, permission_level="manage"))
        db.add(UserProjectAccess(user_id=users[1].id, project_id=proj_research.id, permission_level="manage"))
        db.add(UserProjectAccess(user_id=users[2].id, project_id=proj_content.id, permission_level="manage"))
        await db.flush()

        # Platform keys (fake encrypted keys for demo)
        keys_data = [
            (proj_chatbot.id, "openai", "OpenAI Chatbot Key"),
            (proj_chatbot.id, "anthropic", "Anthropic Chatbot Key"),
            (proj_internal.id, "openai", "OpenAI Internal Key"),
            (proj_research.id, "anthropic", "Anthropic Research Key"),
            (proj_research.id, "google", "Gemini Research Key"),
            (proj_content.id, "openai", "OpenAI Content Key"),
            (proj_content.id, "google", "Gemini Content Key"),
        ]
        platform_keys = []
        for project_id, platform, label in keys_data:
            pk = PlatformKey(
                project_id=project_id,
                platform=platform,
                api_key_encrypted=encrypt_api_key(f"demo-{platform}-key-{uuid.uuid4().hex[:8]}"),
                key_label=label,
                last_synced_at=datetime.now(timezone.utc),
            )
            db.add(pk)
            platform_keys.append(pk)
        await db.flush()

        # Generate usage records for the last 60 days
        today = date.today()
        for pk in platform_keys:
            models = MODELS_BY_PLATFORM.get(pk.platform, ["unknown"])
            for day_offset in range(60):
                d = today - timedelta(days=day_offset)
                for model in random.sample(models, min(len(models), random.randint(1, 3))):
                    input_tok = random.randint(500, 50000)
                    output_tok = random.randint(200, 20000)
                    total_tok = input_tok + output_tok

                    cost_per_1k = {"gpt-4o": 0.005, "gpt-4o-mini": 0.0003, "gpt-4-turbo": 0.01, "gpt-3.5-turbo": 0.001,
                                   "text-embedding-3-small": 0.00002, "claude-sonnet-4-20250514": 0.003, "claude-3.5-haiku": 0.001,
                                   "claude-3-opus": 0.015, "gemini-2.0-flash": 0.0001, "gemini-1.5-pro": 0.00125,
                                   "gemini-1.5-flash": 0.000075}.get(model, 0.001)
                    cost = Decimal(str(round(total_tok * cost_per_1k / 1000, 6)))
                    requests = random.randint(10, 500)

                    db.add(UsageRecord(
                        platform_key_id=pk.id, recorded_at=d, model=model,
                        input_tokens=input_tok, output_tokens=output_tok, total_tokens=total_tok,
                        cost_usd=cost, request_count=requests,
                    ))

        await db.commit()
        print("Seed complete!")
        print(f"  Admin login: admin@demo.com / admin123")
        print(f"  User logins: alice@demo.com, bob@demo.com, carol@demo.com / user123")
        print(f"  Tenant: {tenant.name} ({tenant.id})")
        print(f"  Departments: 3, Projects: 4, Keys: 7")
        print(f"  ~60 days of usage data generated")


if __name__ == "__main__":
    asyncio.run(seed())

File: prisma/seed.ts
import { PrismaClient } from "@prisma/client";
import { hashSync } from "bcryptjs";

const prisma = new PrismaClient();

function rand(min: number, max: number): number {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

function randFloat(min: number, max: number): number {
  return Math.round((Math.random() * (max - min) + min) * 10000) / 10000;
}

const MODEL_CONFIG: Record<string, { platform: string; inputCostPer1k: number; outputCostPer1k: number }> = {
  "gpt-4o": { platform: "openai", inputCostPer1k: 0.005, outputCostPer1k: 0.015 },
  "gpt-4o-mini": { platform: "openai", inputCostPer1k: 0.00015, outputCostPer1k: 0.0006 },
  "gpt-4-turbo": { platform: "openai", inputCostPer1k: 0.01, outputCostPer1k: 0.03 },
  "gpt-3.5-turbo": { platform: "openai", inputCostPer1k: 0.0005, outputCostPer1k: 0.0015 },
  "text-embedding-3-small": { platform: "openai", inputCostPer1k: 0.00002, outputCostPer1k: 0 },
  "claude-sonnet-4-20250514": { platform: "anthropic", inputCostPer1k: 0.003, outputCostPer1k: 0.015 },
  "claude-3.5-haiku-20241022": { platform: "anthropic", inputCostPer1k: 0.0008, outputCostPer1k: 0.004 },
  "claude-3-opus-20240229": { platform: "anthropic", inputCostPer1k: 0.015, outputCostPer1k: 0.075 },
  "gemini-2.0-flash": { platform: "google", inputCostPer1k: 0.0001, outputCostPer1k: 0.0004 },
  "gemini-1.5-pro": { platform: "google", inputCostPer1k: 0.00125, outputCostPer1k: 0.005 },
};

interface KeyDef {
  platform: string;
  keyLabel: string;
  models: string[];
}

async function main() {
  console.log("Seeding database with mock data...\n");

  // Clear existing usage data for idempotent re-seeding
  await prisma.usageRecord.deleteMany();
  await prisma.billingSnapshot.deleteMany();
  await prisma.syncLog.deleteMany();
  await prisma.platformKey.deleteMany();
  await prisma.adminKey.deleteMany();
  await prisma.userProjectAccess.deleteMany();

  const tenant = await prisma.tenant.upsert({
    where: { slug: "demo-org" },
    update: {},
    create: { name: "Demo Organization", slug: "demo-org" },
  });

  // Departments
  const engineering = await prisma.department.upsert({
    where: { tenantId_name: { tenantId: tenant.id, name: "Engineering" } },
    update: {},
    create: { tenantId: tenant.id, name: "Engineering", description: "Platform & backend engineering" },
  });
  const ml = await prisma.department.upsert({
    where: { tenantId_name: { tenantId: tenant.id, name: "Machine Learning" } },
    update: {},
    create: { tenantId: tenant.id, name: "Machine Learning", description: "ML research and fine-tuning" },
  });
  const product = await prisma.department.upsert({
    where: { tenantId_name: { tenantId: tenant.id, name: "Product" } },
    update: {},
    create: { tenantId: tenant.id, name: "Product", description: "Product development and design" },
  });
  const marketing = await prisma.department.upsert({
    where: { tenantId_name: { tenantId: tenant.id, name: "Marketing" } },
    update: {},
    create: { tenantId: tenant.id, name: "Marketing", description: "Content and growth marketing" },
  });

  // Users
  const admin = await prisma.user.upsert({
    where: { email: "admin@demo.com" },
    update: {},
    create: { tenantId: tenant.id, email: "admin@demo.com", passwordHash: hashSync("admin123", 12), fullName: "Admin User", role: "admin", departmentId: engineering.id },
  });
  const alice = await prisma.user.upsert({
    where: { email: "alice@demo.com" },
    update: {},
    create: { tenantId: tenant.id, email: "alice@demo.com", passwordHash: hashSync("user123", 12), fullName: "Alice Chen", role: "user", departmentId: engineering.id },
  });
  const bob = await prisma.user.upsert({
    where: { email: "bob@demo.com" },
    update: {},
    create: { tenantId: tenant.id, email: "bob@demo.com", passwordHash: hashSync("user123", 12), fullName: "Bob Kumar", role: "user", departmentId: ml.id },
  });
  const carol = await prisma.user.upsert({
    where: { email: "carol@demo.com" },
    update: {},
    create: { tenantId: tenant.id, email: "carol@demo.com", passwordHash: hashSync("user123", 12), fullName: "Carol Martinez", role: "user", departmentId: product.id },
  });
  const dave = await prisma.user.upsert({
    where: { email: "dave@demo.com" },
    update: {},
    create: { tenantId: tenant.id, email: "dave@demo.com", passwordHash: hashSync("user123", 12), fullName: "Dave Thompson", role: "user", departmentId: marketing.id },
  });

  // Projects
  const chatbot = await prisma.project.upsert({
    where: { tenantId_name: { tenantId: tenant.id, name: "Chatbot Production" } },
    update: {},
    create: { tenantId: tenant.id, name: "Chatbot Production", departmentId: engineering.id, description: "Customer-facing chatbot" },
  });
  const internalTools = await prisma.project.upsert({
    where: { tenantId_name: { tenantId: tenant.id, name: "Internal Tools" } },
    update: {},
    create: { tenantId: tenant.id, name: "Internal Tools", departmentId: engineering.id, description: "Internal developer tools" },
  });
  const research = await prisma.project.upsert({
    where: { tenantId_name: { tenantId: tenant.id, name: "ML Research" } },
    update: {},
    create: { tenantId: tenant.id, name: "ML Research", departmentId: ml.id, description: "Model evaluation and benchmarking" },
  });
  const contentGen = await prisma.project.upsert({
    where: { tenantId_name: { tenantId: tenant.id, name: "Content Generation" } },
    update: {},
    create: { tenantId: tenant.id, name: "Content Generation", departmentId: product.id, description: "Blog posts, docs, marketing copy" },
  });
  const embeddings = await prisma.project.upsert({
    where: { tenantId_name: { tenantId: tenant.id, name: "Search & Embeddings" } },
    update: {},
    create: { tenantId: tenant.id, name: "Search & Embeddings", departmentId: engineering.id, description: "Semantic search pipeline" },
  });

  // User access
  const accessPairs = [
    { userId: alice.id, projectId: chatbot.id, permissionLevel: "manage" },
    { userId: alice.id, projectId: internalTools.id, permissionLevel: "manage" },
    { userId: bob.id, projectId: research.id, permissionLevel: "manage" },
    { userId: bob.id, projectId: embeddings.id, permissionLevel: "view" },
    { userId: carol.id, projectId: contentGen.id, permissionLevel: "manage" },
    { userId: carol.id, projectId: chatbot.id, permissionLevel: "view" },
    { userId: dave.id, projectId: contentGen.id, permissionLevel: "view" },
  ];
  for (const a of accessPairs) {
    await prisma.userProjectAccess.upsert({
      where: { userId_projectId: { userId: a.userId, projectId: a.projectId } },
      update: {},
      create: a,
    });
  }

  // Platform keys per project
  const keyDefs: { projectId: string; keys: KeyDef[] }[] = [
    {
      projectId: chatbot.id,
      keys: [
        { platform: "openai", keyLabel: "Chatbot GPT-4o Key", models: ["gpt-4o", "gpt-4o-mini"] },
        { platform: "anthropic", keyLabel: "Chatbot Claude Key", models: ["claude-sonnet-4-20250514", "claude-3.5-haiku-20241022"] },
      ],
    },
    {
      projectId: internalTools.id,
      keys: [
        { platform: "openai", keyLabel: "Internal Tools GPT Key", models: ["gpt-4o-mini", "gpt-3.5-turbo"] },
        { platform: "google", keyLabel: "Internal Gemini Key", models: ["gemini-2.0-flash"] },
      ],
    },
    {
      projectId: research.id,
      keys: [
        { platform: "openai", keyLabel: "Research GPT-4 Turbo", models: ["gpt-4-turbo", "gpt-4o"] },
        { platform: "anthropic", keyLabel: "Research Claude Opus", models: ["claude-3-opus-20240229", "claude-sonnet-4-20250514"] },
        { platform: "google", keyLabel: "Research Gemini Pro", models: ["gemini-1.5-pro", "gemini-2.0-flash"] },
      ],
    },
    {
      projectId: contentGen.id,
      keys: [
        { platform: "openai", keyLabel: "Content GPT-4o Key", models: ["gpt-4o", "gpt-4o-mini"] },
        { platform: "anthropic", keyLabel: "Content Claude Key", models: ["claude-sonnet-4-20250514"] },
      ],
    },
    {
      projectId: embeddings.id,
      keys: [
        { platform: "openai", keyLabel: "Embeddings Key", models: ["text-embedding-3-small"] },
      ],
    },
  ];

  const DAYS = 60;
  const now = new Date();
  let totalRecords = 0;

  for (const { projectId, keys } of keyDefs) {
    for (const keyDef of keys) {
      const pk = await prisma.platformKey.create({
        data: {
          projectId,
          platform: keyDef.platform,
          keyLabel: keyDef.keyLabel,
          redactedValue: `...${rand(1000, 9999)}`,
          isActive: true,
        },
      });

      const usageRecords = [];

      for (let d = 0; d < DAYS; d++) {
        const date = new Date(now);
        date.setDate(date.getDate() - d);
        date.setHours(0, 0, 0, 0);

        // Weekday multiplier (less usage on weekends)
        const dayOfWeek = date.getDay();
        const weekdayMult = dayOfWeek === 0 || dayOfWeek === 6 ? 0.3 : 1;

        // Gradual growth trend (more recent = more usage)
        const growthMult = 0.6 + (0.4 * (DAYS - d) / DAYS);

        for (const model of keyDef.models) {
          const config = MODEL_CONFIG[model];
          const isEmbedding = model.includes("embedding");

          const baseInput = isEmbedding ? rand(50000, 500000) : rand(2000, 80000);
          const baseOutput = isEmbedding ? 0 : rand(500, 30000);

          const inputTokens = Math.round(baseInput * weekdayMult * growthMult);
          const outputTokens = Math.round(baseOutput * weekdayMult * growthMult);
          const totalTokens = inputTokens + outputTokens;
          const costUsd = (inputTokens * config.inputCostPer1k + outputTokens * config.outputCostPer1k) / 1000;
          const requestCount = isEmbedding ? rand(100, 2000) : rand(10, 500);

          const userPool = ["alice.chen", "bob.kumar", "carol.martinez", "dave.thompson", "system-bot"];
          usageRecords.push({
            platformKeyId: pk.id,
            recordedAt: date,
            model,
            inputTokens,
            outputTokens,
            totalTokens,
            costUsd: Math.round(costUsd * 1000000) / 1000000,
            requestCount: Math.round(requestCount * weekdayMult * growthMult),
            userIdExternal: userPool[rand(0, userPool.length - 1)],
          });
        }
      }

      await prisma.usageRecord.createMany({ data: usageRecords });
      totalRecords += usageRecords.length;

      // Billing snapshot for last 30 days
      const last30 = usageRecords.filter((r) => {
        const diff = (now.getTime() - r.recordedAt.getTime()) / (1000 * 60 * 60 * 24);
        return diff <= 30;
      });
      const totalCost30 = last30.reduce((sum, r) => sum + r.costUsd, 0);
      const totalTokens30 = last30.reduce((sum, r) => sum + r.totalTokens, 0);

      await prisma.billingSnapshot.create({
        data: {
          platformKeyId: pk.id,
          periodStart: new Date(now.getTime() - 30 * 24 * 60 * 60 * 1000),
          periodEnd: now,
          totalCostUsd: Math.round(totalCost30 * 100) / 100,
          totalTokens: totalTokens30,
          breakdown: JSON.parse(JSON.stringify(
            last30.reduce((acc, r) => {
              acc[r.model] = (acc[r.model] || 0) + r.costUsd;
              return acc;
            }, {} as Record<string, number>)
          )),
        },
      });

      // Sync log
      await prisma.syncLog.create({
        data: {
          platformKeyId: pk.id,
          syncType: "auto",
          status: "success",
          recordsSynced: usageRecords.length,
          completedAt: now,
        },
      });

      console.log(`  ${keyDef.keyLabel}: ${usageRecords.length} usage records`);
    }
  }

  console.log(`\nTotal usage records created: ${totalRecords}`);
  console.log(`\nSeed complete!`);
  console.log(`\nLogin credentials:`);
  console.log(`  Admin:  admin@demo.com / admin123`);
  console.log(`  Users:  alice@demo.com / user123`);
  console.log(`          bob@demo.com   / user123`);
  console.log(`          carol@demo.com / user123`);
  console.log(`          dave@demo.com  / user123`);
}

main()
  .catch((e) => {
    console.error(e);
    process.exit(1);
  })
  .finally(() => prisma.$disconnect());
